### PR TITLE
feat: additional Array.prototype.split() safety

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

follows-up on https://github.com/shapeshift/web/pull/6810 - since we can't do the `noUncheckedIndexedAccess` dance without having 500+ errors, this at least brings strictness for the specific
case of `Array.prototype.split()` returning `(string | undefined)[]` vs. `string` with our current compiler options.

Note that this aims to be a non-breaking change - while the strictness is now present, all type violations have been tackled in a way that should be non-breaking, with either:

- nullish coalescing to `''` (i.e no different than before, since we were assuming a string)
- throwing exceptions when applicable, for the specific cases where this doesn't change the flow and things would already fail as-is

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

N/A, follows-up on https://github.com/shapeshift/web/pull/6810 

> What protocols, transaction types or contract interactions might be affected by this PR?

Low but not nil by definition - see testing steps

## Testing

- THOR swaps are still happy with no error thrown and with slippage still applied in the memo as before

The two above don't require new Tx - just do a full cache clear and ensure Txs look sane against develop

- THOR Txs with a memo are still parsed in Tx history without errors thrown (e.g swaps)
- THOR Txs without a memo are still parsed in Tx history without errors thrown (e.g sends)

- Wallet creation is still happy on mobile (PR deployed to gome, just point your mobile app to gome and test it out)
- Backup passphrase is still happy (test on either desktop or mobile)

### Engineering

- ☝🏽 
- Confirm the non-breaking change nature of this conceptually

### Operations

- ☝🏽 

### Txs look sane after a cache clear (left: develop, right: this diff with cleared cache)

 
image

- THOR swaps are still happy with no error thrown and with slippage still applied in the memo as before

image
image

- Wallet creation is still happy on mobile (tested on linked upstream gome in the Android app)

  - Wrong word
  image
 - Correct words
 
image
image

- Backup passphrase is still happy

image
image

### CODE REVIEW

File: src/lib/globals.ts
Comment (gomesalexandre) [inline src/lib/globals.ts:12]: The actual safety

File: scripts/generateTradableThorAssetMap/utils.ts
Comment (gomesalexandre) [inline scripts/generateTradableThorAssetMap/utils.ts:110]: i.e non-breaking change, we were assuming a string previously

File: src/components/Layout/Header/NavBar/Native/BackupPassphraseModal/BackupPassphraseTest.tsx
Comment (gomesalexandre) [inline src/components/Layout/Header/NavBar/Native/BackupPassphraseModal/BackupPassphraseTest.tsx:70]: i.e non-breaking change, we were assuming a string previously

File: src/context/WalletProvider/MobileWallet/components/MobileCreateTest.tsx
Comment (gomesalexandre) [inline src/context/WalletProvider/MobileWallet/components/MobileCreateTest.tsx:71]: i.e non-breaking change, we were assuming a string previously

File: src/lib/swapper/swappers/ThorchainSwapper/utils/makeSwapMemo/assertIsValidMemo.ts
Comment (gomesalexandre) [inline src/lib/swapper/swappers/ThorchainSwapper/utils/makeSwapMemo/assertIsValidMemo.ts:72]: Non-breaking change - the line below would throw anyway if we were trying to call `.split()` on `undefined`
image

File: src/lib/swapper/swappers/ThorchainSwapper/utils/makeSwapMemo/assertIsValidMemo.ts
Comment (gomesalexandre) [inline src/lib/swapper/swappers/ThorchainSwapper/utils/makeSwapMemo/assertIsValidMemo.ts:74]: Non-breaking change, we should always have a valid pool (i.e consisting of chain and asset parts).
Assuming we didn't on some calls:

- previous l.72 would evalute to undefined if `pool.split(POOL_PART_DELIMITER)[0]` is 
- `isAddressValid` would then be undefined 
- we would throw right under either way

File: src/lib/utils/index.ts
Comment (gomesalexandre) [inline src/lib/utils/index.ts:180]: i.e non-breaking change, we were assuming a string previously

File: src/state/slices/txHistorySlice/utils.ts
Comment (gomesalexandre) [inline src/state/slices/txHistorySlice/utils.ts:67]: Technically a breaking change in a way, a non-breaking one would be to have non-nullish coallescing to `''`,  though if some or all were ever undefined, we would upsert Txs with parts being undefined in the store, and then, problems.

File: src/state/slices/txHistorySlice/utils.ts
Comment (gomesalexandre) [inline src/state/slices/txHistorySlice/utils.ts:84]: Ditto here - technically a breaking change in away but if we ever had 4 parts but no memo, problems.

File: src/lib/swapper/swappers/ThorchainSwapper/utils/makeSwapMemo/assertIsValidMemo.ts
Comment (gomesalexandre) [inline src/lib/swapper/swappers/ThorchainSwapper/utils/makeSwapMemo/assertIsValidMemo.ts:1]: Also worth noting this goes away soon in profit of `assertAndProcessMemo`, with the validation bits moved there

File: src/lib/swapper/swappers/ThorchainSwapper/utils/makeSwapMemo/assertIsValidMemo.ts
Comment (woodenfurniture) [reply src/lib/swapper/swappers/ThorchainSwapper/utils/makeSwapMemo/assertIsValidMemo.ts:72]: The root cause here is we're missing `noUncheckedIndexedAccess` in our `tsconfig.json`. Might be worth checking the work needed to enable this as it'd prevent most cases but likely has thousands of violations

File: src/lib/globals.ts
Comment (woodenfurniture) [reply src/lib/globals.ts:12]: See https://github.com/shapeshift/web/pull/6811/files#r1591577558 , we can use typescript for this.

We likely have many occurrences where we're accessing arrays with missing values, not just in the case of `String.split()`, but the sheer number of them might mean we're best off starting here and following up with the blanket rule.

Comment (woodenfurniture) [reply src/lib/globals.ts:12]: Ah, I see your original comment

> since we can't do the `noUncheckedIndexedAccess` dance without having 500+ errors, this at least brings strictness for the specific

File: src/lib/swapper/swappers/ThorchainSwapper/utils/makeSwapMemo/assertIsValidMemo.ts
Comment (woodenfurniture) [reply src/lib/swapper/swappers/ThorchainSwapper/utils/makeSwapMemo/assertIsValidMemo.ts:72]: Ah, I see your original comment

> since we can't do the `noUncheckedIndexedAccess` dance without having 500+ errors, this at least brings strictness for the specific

File: src/context/WalletProvider/NativeWallet/components/NativeTestPhrase.tsx
Comment (woodenfurniture) [inline src/context/WalletProvider/NativeWallet/components/NativeTestPhrase.tsx:52]: 👀

File: src/lib/globals.ts
Comment (woodenfurniture) [reply src/lib/globals.ts:12]: 🧠

File: src/state/slices/txHistorySlice/utils.ts
Comment (woodenfurniture) [reply src/state/slices/txHistorySlice/utils.ts:67]: Heh, breaking as in handling the error in the correct place 🧠

File: src/pages/ThorChainLP/utils.ts
Comment (gomesalexandre) [reply src/pages/ThorChainLP/utils.ts:12]: How about 5b8e3ed3fe0f503a73fd79dbc8dc10e9538135e

Comment (woodenfurniture) [reply src/pages/ThorChainLP/utils.ts:12]: goat

**Review by woodenfurniture — APPROVED** [review]
Needs rebase but LGTM after that

**Pull request merged**
